### PR TITLE
fix(python,rust): Allow append/zip_with/extend on local categoricals

### LINKED ISSUE #10393: Add `synchronize_categoricals()`
### Problem description

I encounter this situation often enough that I feel it may be useful to others: you have `df1` with some categoricals that were generated separately from `df2` and thus use disparate string caches. The following example shows what you currently have to do to get them to play nicely:

```python
import polars as pl
from polars import col

df1 = pl.DataFrame({
    'a': pl.Series(['a', 'b', 'c'], dtype=pl.Categorical)
})
df2 = pl.DataFrame({
    'a': pl.Series(['d', 'e', 'f'], dtype=pl.Categorical)
})

df1.vstack(df2) # error!! df1 and df2 use different string caches
```
If you get `df1` and `df2` from separate processes, your only recourse is the following:

```python
with pl.StringCache():
    df1 = df1.with_columns(col('a').cast(pl.Utf8).cast(pl.Categorical))
    df2 = df2.with_columns(col('a').cast(pl.Utf8).cast(pl.Categorical))

df1.vstack(df2) # works!
```

This seems a bit wasteful to do a full conversion. It would be nice to able to do:

```python
pl.synchronize_categoricals((df1, df2, ...))
```
...or something similar.

### LINKED ISSUE #12091: Incorrect merge of categorical columns if they overlap
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
s1 = pl.Series(['a','b','c'],dtype=pl.Categorical)
s2 = pl.Series(['b','c','d'],dtype=pl.Categorical)
s1.extend(s2)
```

### Log output

```shell
shape: (6,)
Series: '' [cat]
[
	"a"
	"b"
	"c"
	"a"
	"b"
	"c"
]
```

### Issue description

For operations that combine two local categoricals (`zip_with`, `append`, `extend`) we incorrectly merge the two types. Right now we do a simple concat of both categories and bump the underlying physical (or in case of `extend` do not change the physical at all). When the categories overlap, this leads to strange results where there are duplicates in the categories list.

### Expected behavior

```
shape: (6,)
Series: '' [cat]
[
	"a"
	"b"
	"c"
	"a"
	"b"
	"d"
]
```

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.12
Index type:          UInt32
Platform:            Linux-6.2.0-35-generic-x86_64-with-glibc2.35
Python:              3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_sqlite:  0.7.0
cloudpickle:         2.2.1
connectorx:          0.3.2
deltalake:           0.11.0
fsspec:              2023.9.2
gevent:              <not installed>
matplotlib:          3.7.2
numpy:               1.24.4
openpyxl:            <not installed>
pandas:              2.0.3
pyarrow:             12.0.0
pydantic:            2.4.2
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          2.0.21
xlsx2csv:            0.8.1
xlsxwriter:          3.1.6```

</details>

### PR DESCRIPTION
**Pull request by c-peters**
fix #12091
fix #10393

### CODE REVIEW

File: crates/polars-core/src/series/implementations/categorical.rs [hunk 1]
```diff
@@ -165,13 +165,8 @@ impl SeriesTrait for SeriesWrap<CategoricalChunked> {
...
     fn extend(&mut self, other: &Series) -> PolarsResult<()> {
         polars_ensure!(self.0.dtype() == other.dtype(), extend);
-        let other = other.categorical()?;
-        self.0.physical_mut().extend(other.physical());
-        let new_rev_map = self.0._merge_categorical_map(other)?;
-        // SAFETY
-        // rev_maps are merged
-        unsafe { self.0.set_rev_map(new_rev_map, false) };
-        Ok(())
+        // TODO: actually implement extend here
```
Comment (ritchie46) [inline crates/polars-core/src/series/implementations/categorical.rs:168]: You can check the rev-map. If they are both from the same global we can keep the old code under a branch.

File: crates/polars-core/src/chunked_array/logical/categorical/merge.rs [hunk 2]
```diff
@@ -102,52 +99,95 @@ impl RevMapMerger {
...
+        } else {
+            idx_mapping.insert(idx as u32, new_categories.len() as u32);
+            new_categories.push(Some(s));
+        }
     }
+    let new_rev_map = Arc::new(RevMapping::Local(new_categories.into()));
+    Ok((
+        ca_right
+            .physical()
+            .into_iter()
```
Comment (ritchie46) [inline crates/polars-core/src/chunked_array/logical/categorical/merge.rs:145]: Can you replace `into_iter` with:


```rust
                self.downcast_iter()
                    .flatten()
```

That ensure we don't go via a `dyn Iterator`.

File: crates/polars-core/src/chunked_array/builder/list/categorical.rs [hunk 3]
```diff
@@ -1,11 +1,121 @@
...
+        let DataType::Categorical(Some(rev_map)) = s.dtype() else {
+            polars_bail!(ComputeError: "expected categorical type")
+        };
+        let RevMapping::Local(cats_right) = &**rev_map else {
+            polars_bail!(string_cache_mismatch)
+        };
+        let ca = s.categorical()?;
+
+        // Map the physical of the appended series to be compatible with the existing rev map
+        let mut idx_mapping = PlHashMap::with_capacity(ca.len());
```
[the remark this answers is not in the document]
Comment (c-peters) [reply crates/polars-core/src/chunked_array/builder/list/categorical.rs:58]: Will do in a follow up PR

**Comment by s-banach** [comment]
Sorry to bother you! Just wondering whether there's a path from here to joining on local categoricals.

**Review by ritchie46 — APPROVED** [review]

**Comment by deanm0000** [comment]
> Sorry to bother you! Just wondering whether there's a path from here to joining on local categoricals.

In the interim, pyarrow will join tables by categorical

```
pl.from_arrow(
df1.to_arrow().join(
df2.to_arrow(), keys='cat_col')
)
```

**Comment by c-peters** [comment]
Yes, in the future it would be possible to join on local categoricals with this merge operation.

**Pull request merged**